lib/app_jwt/app_jwt.c

```diff
@@ -0,0 +1,767 @@
+	hdr_b64 = convert_str_to_b64_url(hdr_str);
+
+	cJSON_free(hdr_str);
+	hdr_str = NULL;
+
+	if (!hdr_b64) {
+		LOG_ERR("Failed to convert header string to base64");
+		return NULL;
+	}
+
+	/* Get expiration time stamp */
+	if (jwt->validity_s > 0) {
+		int err = clock_gettime(CLOCK_REALTIME, &tp);
+
+		if (err) {
+			/* clock_gettime error, use 0 value */
+			exp_ts_s = 0;
+		} else {
+			exp_ts_s = tp.tv_sec;
+		}
+
+		exp_ts_s += jwt->validity_s;
+	}
+
+	/* Create the payload */
+	pay_str = jwt_payload_create(jwt->subject, jwt->audience, exp_ts_s);
+	if (!pay_str) {
+		LOG_ERR("Failed to create JWT JSON payload");
+		return NULL;
+	}
+
+	/* Convert payload JSON string to base64 URL */
+	pay_b64 = convert_str_to_b64_url(pay_str);
+
+	cJSON_free(pay_str);
+	pay_str = NULL;
+
+	if (!pay_b64) {
+		LOG_ERR("Failed to convert payload string to base64");
+		return NULL;
+	}
+
+	/* Create the base64 URL data to be signed */
+	unsigned_jwt = jwt_header_payload_combine(hdr_b64, pay_b64);
+
+	free(hdr_b64);
+	hdr_b64 = NULL;
+
+	free(pay_b64);
+	pay_b64 = NULL;
+
+	return unsigned_jwt;
+}
+
+static int jwt_signature_get(const uint32_t user_key_id, const int sec_tag, const char *const jwt,
+			     char *const sig_buf, size_t sig_sz)
+{
+	int err;
+
+	size_t o_len;
+
+	uint8_t sig_raw[ECDSA_SHA_256_SIG_SZ];
+
+	/* Use Application IAK key for signing the JWT */
+	/* ignore the provided key_id and sec_tag */
+	(void)sec_tag;
+	(void)user_key_id;
+	err = sign_message(IAK_APPLICATION_GEN1, jwt, strlen(jwt), sig_raw, ECDSA_SHA_256_SIG_SZ,
+			   &o_len);
+	if (err) {
+		LOG_ERR("Failed to sign message : %d", err);
+		return -EACCES;
+	}
+
+#if defined(CONFIG_APP_JWT_VERIFY_SIGNATURE)
+	err = verify_message_signature(IAK_APPLICATION_GEN1, jwt, strlen(jwt), sig_raw, o_len);
+	if (err) {
+		LOG_ERR("Failed to verify message : %d", err);
+		return -EACCES;
+	}
+#endif /* CONFIG_APP_JWT_VERIFY_SIGNATURE */
+
+	/* Convert signature to base64 URL */
+	err = base64_encode(sig_buf, sig_sz, &o_len, sig_raw, sizeof(sig_raw));
+	if (err) {
+		LOG_ERR("base64_encode failed, error: %d", err);
+		return -EIO;
+	}
+
+	base64_url_format(sig_buf);
+	return 0;
+}
+
+static int jwt_signature_append(const char *const unsigned_jwt, const char *const sig,
+				char *const jwt_buf, size_t jwt_sz)
+{
+	int err = 0;
+
+	/* Get the size of the final, signed JWT: +1 for */
+	/* '.' and null-terminator */
+	size_t final_sz = strlen(unsigned_jwt) + 1 + strlen(sig) + 1;
+
+	if (final_sz > jwt_sz) {
+		/* Provided buffer is too small */
+		err = -E2BIG;
+	}
+
+	/* JWT final form:
+	 * <base64_header>.<base64_payload>.<base64_signature>
+	 */
+	if (err == 0) {
+		int ret = snprintk(jwt_buf, jwt_sz, "%s.%s", unsigned_jwt, sig);
+
+		if ((ret < 0) || (ret >= jwt_sz)) {
+			err = -ETXTBSY;
+		}
+	}
+
+	return err;
+}
+
+int app_jwt_generate(struct app_jwt_data *const jwt)
+{
+	if (jwt == NULL) {
+		return -EINVAL;
+	}
+
+	if ((jwt->jwt_buf) && (jwt->jwt_sz == 0)) {
```

Comment:
I'm not sure what this tries to do. It would make sense to check that `jwt_buf != NULL` and `jwt_sz != 0`, but that's not what this currently does.

Reply:
in case the buffer is valid and the token size is set to 0 there should be an error.
this was done in order to make sure this unit test from "tests/lib/modem_jwt/src/jwt_test.c" passes : 

https://github.com/nrfconnect/sdk-nrf/blob/541f7bf084c2f9c37e1e1fc8e1dbd3ab0c1538cf/tests/lib/modem_jwt/src/jwt_test.c#L61

note : I didn't run the unit tests, I just tried to keep the expected error cases to be still applicable.

Reply:
Wouldn't this be better? The pointer should be valid and the size should be greater than zero, right?
```suggestion
	if ((jwt->jwt_buf == NULL) || (jwt->jwt_sz == 0)) {
```